### ISSUE TITLE
game spams message -- ship does not have enough crew

### ISSUE THREAD

**Issue by starpengiun** [endless-sky/endless-sky#8663]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Greetings,

I have run into a problem regarding  the message scroll at the bottom of the screen.  It is spamming the following message:

"The <ship name> is moving erratically because there are not enough crew to pilot it". 

I checked my fleet's roster and none of the ships in the messages are in my fleet.  These seem to be AI-generated ships.  

The spammed message problem has reached the point it is preventing me from seeing important information.  For example, so many messages are scrolling past regarding the lack of crew space on other ships that I have missed seeing the messages of my ships being destroyed.  Given the fact the destruction message is in a different color than standard messages and I still missed it, that should give you an idea on how fast these messages are cropping up and scrolling.  

Possibly related, just letting my ships sit in orbit is putting the in-game CPU calculator over 100%, so I suspect the continued spamming of messages is stressing the processor a bit.  My system monitor notes the game is using 40% or so, which seems high as well.  This is in the screenshot below.

The spamming is getting worse the longer the game runs, probably because of the generating of new ships.

I would hate to remove the mods I am using (the second screen shot below), but maybe one or more of them are the cause?  

I am uncertain where to go at this point.  I am playing the game, but not yet attempted to make any mods.  Any help is appreciated.  

I hope I did the screenshots right.  This is the first time I have created one in this OS system.  I also hope I did not make too big of a mess with this posting.  

![spam ship message](https://user-images.githubusercontent.com/129539323/232255802-9020da05-df81-4979-81ad-c76fad4b1383.png)

### Steps to Reproduce

2. Click on '...'
3. Scroll down to '...'

### Expected Behavior

The CPU not running 100%+, causing frame rates to slow down.

The removal of the message for all ships but my own in my fleet.  I do not think I need to know about the problems with other fleets, captains, and crews :-)  

### Screenshots

![endless sky modos list](https://user-images.githubusercontent.com/129539323/232256164-1a16cf4c-d506-4434-9800-4f5005d547c3.png)

### Operating System

Zorin OS 16

### Game Source

GitHub Releases

### Game Version

V10.0

### Additional Information

I searched for answers to my problem and I think it is related to issue #1079 and maybe #1095.  Those threads appear to be closed, so I was reluctant to post there.  Some places, after they close a thread, auto-ban the account.  Since I did not know the rules on that here, I decided to post this as a new issue.  If it needs to be moved, please let me know where it ends up.  

I am running a group of mods within the game (screenshot below), and I think I read somewhere in the issues that if the weapon capacity or output capacity are negative numbers, then there might be issues?  I did run into a couple of pirate quicksilver vessels that had 9 crew in 7 bunks, which created an issue (solved by selling the ships), but that was just the singular instance where I captured a ship (both cases a quicksilver).  I did not think at the time to check the name to see if it matched one of the spammed names.  

Thanks in advance for any help.

**Comment by starpengiun** [comment]
Sorry, I meant to say the CPU is running over 100%.  Sorry about the typo

**Comment by warp-core** [comment]
Can you upload your save file?

**Comment by starpengiun** [comment]
[Perseus Starseeker~perseus-spam-messages.txt](https://github.com/endless-sky/endless-sky/files/11240921/Perseus.Starseeker.perseus-spam-messages.txt)

Oops.  Sorry.  My mistake.

**Comment by warp-core** [comment]
I suspect the high CPU usage is due to the very large number of NPC ships from many (plugin) bounty missions you have, however I cannot properly assess the impact of this myself because I do not have all those plugins and so many of these missions are disabled.
As for the ships lacking crew, have you attempted to capture many ships but abandoned them before succeeding, leaving them with insufficient crew?

**Comment by starpengiun** [comment]
Hmmm... 

Okay, to the CPU usage, if I were to abort a bunch of the uncompleted missions, what would that do?  Do you think that would help with the decline of the CPU usage?  Maybe impose a self-limit of active missions until I get a better idea on what the CPU can handle?  

Also, you mention a lot of the missions are disabled.  What do you mean, please?  Are they disabled because the main game will not permit them to run due to changes in the core programming?  Or are they disabled in the mod(s) themselves?  Just curious.

As for the ships, I did that with only a few ships, once I realized I did not have a reasonable chance to take it. Usually those ships were destroyed, though, often by me or someone else.  

There were many times I tried to capture a ship, but it was destroyed before I could capture it.  This was usually done by either allies who did not know when to stop shooting, passing merchants ships who felt like being obnoxious,  or other pirates who came into the area to deprive me of the prize.  

However, there have been several ships left adrift, without an attempt to capture, because I was in the middle of a jump and just left them behind or I ran into a large enemy fleet and got out of the area, possibly leaving disabled enemy vessels in  my wake.  

Does that help?

**Comment by warp-core** [comment]
> Okay, to the CPU usage, if I were to abort a bunch of the uncompleted missions, what would that do? Do you think that would help with the decline of the CPU usage? Maybe impose a self-limit of active missions until I get a better idea on what the CPU can handle?

Yes, I expect aborting those missions will reduce CPU usage.

> Also, you mention a lot of the missions are disabled. What do you mean, please? Are they disabled because the main game will not permit them to run due to changes in the core programming? Or are they disabled in the mod(s) themselves? Just curious.

Those missions contain references to some data that the vanilla game does not have, most likely from one of your plugins, but since I don't have that, the game disables the missions for me, meaning they are saved as is, but they are not active in game. Most likely, these missions spawn ships which list outfits that are from plugins.
Because these missions are disabled, their ships are not spawned for me and so cannot contribute to the CPU load.

If you have not left these ships with insufficient crew, I would need to look at the plugins to investigate further. It's possible those make changes that leave these ships with insufficient crew.

**Comment by starpengiun** [comment]
Okay, so, for the CPU usage, I will start deleting unfinished bounty missions and see what that does.  

As to the non-standard references, I am guessing you can find and check the suspected plug-in mods at some point and see what might be going on, since I did find most if not all of them on the list of plug-ins on the endless sky page.  Please let me know when you find a solution for that issue.  I would definitely like ti implement it. 

Thank you very much for your help!  :-)

**Comment by starpengiun** [comment]
Just to let you know, the deletion of several bounty hunter (mod) missions cut the CPU strain down almost by half.  So, anyone who runs into this problem in the future and they are using this mod, self-imposed limits (or, if the mod-maker wants to do it at some point, sent a top-end limit of jobs in the queue) might be the fix they need.

**Comment by petervdmeer** [comment]
I was able to reproduce the issue of ""The [some NPC ship] is moving erratically because there are not enough crew to pilot it" (on a vanilla installation without plugins) with the attached savegame. Changing status to confirmed (although the message is not "spamming", since I only have 4 NPCs in the savegame that have insufficient crew).
[a b.txt](https://github.com/endless-sky/endless-sky/files/11347366/a.b.txt)

**Comment by MiscelaneousItem** [comment]
> I was able to reproduce the issue of "The [some NPC ship] is moving erratically because there are not enough crew to pilot it" (on a vanilla installation without plugins) with the attached savegame. Changing status to confirmed (although the message is not "spamming", since I only have 4 NPCs in the savegame that have insufficient crew). [a b.txt](https://github.com/endless-sky/endless-sky/files/11347366/a.b.txt)

But isn't this just normal behavior ?
What is abnormal is the spamming reported by the op. But that is most likely due to mods.

**Comment by quyykk** [comment]
> But isn't this just normal behavior ?

Yes for player ships, but this seems to be for NPC ships where it is not intended.

**Comment by MiscelaneousItem** [comment]
```
               if(parent.lock() || !isYours)
                        Messages::Add("The " + name + " is moving erratically because there are not enough crew to pilot it."
                                , Messages::Importance::Low);
                else
                        Messages::Add("Your ship is moving erratically because you do not have enough crew to pilot it."
                                , Messages::Importance::Low);
```
From my understanding of `Ship.cpp` and `Ship.h` `isYours` is true for any ship owned by the player. 
So the issues seems to me to be with this condition `if(parent.lock() || !isYours)`

**Comment by starpengiun** [comment]
Luck may have struck in a recent game session regarding this issue, maybe this information will help?

Started a new game, and after some time, I started getting this message again... and then I got lucky.   I ran into the ship generating the message... and managed to capture it!  Yay, a clue!?!?

The ship, as far as I can tell, was generated by a mod (90% certain it was the BetterVanillaPirates mod, because it was part of a fleet related to a bounty hunt).  I parked it on a planet, and then looked at it, and took a picture of what I found:

![Screenshot from 2023-04-28 13-06-27](https://user-images.githubusercontent.com/129539323/235325999-4f45d6a4-216c-4e87-9e42-99e6dbf17326.png)

Says 9 required crew, only 7 bunks.  So, this ship was apparently generated / built with not enough bunks for the crew.  I am guessing the problem is the security station component, which requires 2 people to run it.  

Footnote:  I ended up disowning the ship because (a) I landed on a planet without an outfitter, so I could not get rid of the security station, and (b) a message came up where I did an examination of the ship when I tried to take off and it basically told me not to try because it will crash.  I have never seen that message before, so I do not know if that is from the mod or the game itself.  This is the first (and only) I actually captured one of the ships tagged with this message, so Ii do not know where that message came from.  

I do not know if this helps, but this is what I discovered by accident and luck yesterday.  

RE:  petervdmeer,  MiscelaneousItem

The spamming for me was up to about 30-40+ messages and climbing, but I was / am using several mods.  I was also stacking up bounty hunts, thus improving the possibilities of this problem showing up.  So, my description of message spamming insofar as the unmodified game goes is an unfair description on my part.  If someone wants to change the title of this thread to say something like:

"<NPC Ship Name> does not have enough crew" message showing up in message scroll

to be more accurate, please be my guest.

**Comment by petervdmeer** [comment]
> Says 9 required crew, only 7 bunks. So, this ship was apparently generated / built with not enough bunks for the crew. I am guessing the problem is the security station component, which requires 2 people to run it.

My test-savegame had a similar setup; more required crew than bunks on NPC ships.

> Footnote: I ended up disowning the ship because (a) I landed on a planet without an outfitter, so I could not get rid of the security station, and (b) a message came up where I did an examination of the ship when I tried to take off and it basically told me not to try because it will crash. I have never seen that message before, so I do not know if that is from the mod or the game itself.

This is a message from the vanilla game, I also encountered it when testing for more required crew than bunks on ships owned by me. We might want to make this a separate bug-report (since we can capture this type of ship, but not take off with it on some planets), but this check has been in vanilla for a long time.
It should normally show up on only planets with an outfitter after a player doesn't install sufficient bunks for the crew.

> The spamming for me was up to about 30-40+ messages and climbing, but I was / am using several mods. I was also stacking up bounty hunts, thus improving the possibilities of this problem showing up. So, my description of message spamming insofar as the unmodified game goes is an unfair description on my part. If someone wants to change the title of this thread ..:

No worries, a fix is already in progress. I hope to close this issue soon.

**Comment by petervdmeer** [comment]
fixed in #8715
